File: day14/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var first, last byte
	pairs := map[string]int{}
	rules := map[string]byte{}
	scanner := bufio.NewScanner(file)
	batch1 := true
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			batch1 = false
			continue
		}
		if batch1 {
			for i := 0; i < len(line)-1; i++ {
				p := line[i : i+2]
				if _, ok := pairs[p]; !ok {
					pairs[p] = 0
				}
				pairs[p]++
			}
			first = line[0]
			last = line[len(line)-1]
			continue
		}
		arr := strings.SplitN(line, " -> ", 2)
		if len(arr) != 2 {
			log.Fatalln("Invalid line")
		}
		if len(arr[1]) != 1 {
			log.Fatalln("Invalid line")
		}
		rules[arr[0]] = arr[1][0]
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	for i := 0; i < 10; i++ {
		pairs = processStep(rules, pairs)
	}
	max, min := maxminCount(pairs, first, last)
	fmt.Println("Part 1:", max-min)
	for i := 0; i < 30; i++ {
		pairs = processStep(rules, pairs)
	}
	max, min = maxminCount(pairs, first, last)
	fmt.Println("Part 2:", max-min)
}

func processStep(rules map[string]byte, pairs map[string]int) map[string]int {
	next := map[string]int{}
	for k, v := range pairs {
		p := rules[k]
		p1 := string([]byte{k[0], p})
		p2 := string([]byte{p, k[1]})
		if _, ok := next[p1]; !ok {
			next[p1] = 0
		}
		next[p1] += v
		if _, ok := next[p2]; !ok {
			next[p2] = 0
		}
		next[p2] += v
	}
	return next
}

func maxminCount(pairs map[string]int, first, last byte) (int, int) {
	max := 0
	min := 0
	counts := map[byte]int{
		first: 1,
		last:  1,
	}
	for k, v := range pairs {
		c1 := k[0]
		if _, ok := counts[c1]; !ok {
			counts[c1] = 0
		}
		counts[c1] += v
		c2 := k[1]
		if _, ok := counts[c2]; !ok {
			counts[c2] = 0
		}
		counts[c2] += v
		max = counts[c2] / 2
		min = counts[c2] / 2
	}
	for _, v := range counts {
		if v/2 > max {
			max = v / 2
		}
		if v/2 < min {
			min = v / 2
		}
	}
	return max, min
}

File: day14/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut first = 0;
    let mut last = 0;
    let mut pairs = HashMap::new();
    let mut rules = HashMap::new();
    let mut track_pairs = true;
    for line in reader.lines() {
        let line = line?;
        if line == "" {
            track_pairs = false;
            continue;
        }
        if track_pairs {
            let v = line.as_bytes().into_iter().collect::<Vec<_>>();
            for k in v.windows(2) {
                let (a, b) = match k {
                    &[&a, &b] => (a, b),
                    _ => return Err("Invalid line".into()),
                };
                *pairs.entry((a, b)).or_insert(0) += 1;
            }
            first = *v[0];
            last = *v[v.len() - 1];
            continue;
        }
        let (lhs, rhs) = match line.split_once(" -> ") {
            Some(eqn) => eqn,
            _ => return Err("Invalid line".into()),
        };
        if let [&a, &b] = lhs.as_bytes().into_iter().collect::<Vec<_>>()[..] {
            rules.insert((a, b), rhs.as_bytes()[0]);
        } else {
            return Err("Invalid line".into());
        }
    }

    for _ in 0..10 {
        pairs = process_step(&rules, pairs);
    }
    let (max, min) = find_max_min(&pairs, first, last);
    println!("Part 1: {}", max - min);
    for _ in 0..30 {
        pairs = process_step(&rules, pairs);
    }
    let (max, min) = find_max_min(&pairs, first, last);
    println!("Part 2: {}", max - min);
    Ok(())
}

fn process_step(
    rules: &HashMap<(u8, u8), u8>,
    pairs: HashMap<(u8, u8), usize>,
) -> HashMap<(u8, u8), usize> {
    let mut next = HashMap::new();
    for ((a, b), v) in pairs {
        let c = rules[&(a, b)];
        *next.entry((a, c)).or_insert(0) += v;
        *next.entry((c, b)).or_insert(0) += v;
    }
    next
}

fn find_max_min(pairs: &HashMap<(u8, u8), usize>, first: u8, last: u8) -> (usize, usize) {
    let mut counts = HashMap::from([(first, 1), (last, 1)]);
    let mut max = 0;
    let mut min = 0;
    for (&(a, b), v) in pairs {
        *counts.entry(a).or_insert(0) += v;
        *counts.entry(b).or_insert(0) += v;
        max = counts[&b] / 2;
        min = counts[&b] / 2;
    }
    counts.into_iter().fold((max, min), |(max, min), (_, v)| {
        (
            if v / 2 > max { v / 2 } else { max },
            if v / 2 < min { v / 2 } else { min },
        )
    })
}
